src/lib.rs

```diff
@@ -175,8 +185,58 @@ impl<'a> FMBuilder<'a> {
     /// to `a` then `$2` will refuse to match against `a` (though `$1` will continue to match
     /// against only `a`). Note that ignorable name matches (see [Self::name_matcher_ignore]) are
     /// never subject to distinct name matching. Defaults to `false`.
-    pub fn distinct_name_matching(mut self, yes: bool) -> Self {
-        self.options.distinct_name_matching = yes;
+    #[deprecated(since = "0.3.1", note = "Please use name_matching_processor instead")]
+    pub fn distinct_name_matching(self, yes: bool) -> Self {
+        if yes {
+            self.name_matching_processor(|names| {
+                names.values().collect::<HashSet<_>>().len() == names.len()
+            })
+        } else {
+            self
+        }
+    }
+
+    /// Add a name matching processor: this takes a [HashMap] of `(key, value)` pairs and must
+    /// return `true` if this is a valid set of pairs or false otherwise. Name matching processors
```

Comment:
What are the `key, value` pairs here though?

Reply:
Right, I think I may have cracked it. This function receives the output of the previously applied matchers as a hashmap (they keys are the names, the values are the text the was matched?) and you have the opportunity to inspect it and decide if it's OK for you.

If I understood correctly, what wasn't obvious to me:

 - this requires normal name matchers to be used first, the output of which is piped into the input of `name matching processor`.
 - the name `name matching processor` lead me to think you could "process" (i.e. mutate) something somewhere, but really you just inspect stuff immutably and give a yes/no decision. Would `name matcher validator` be a better name?

 (is the validator called once for each matcher, or with the union of all matchers?)

Reply:
>This function receives the output of the previously applied matchers as a hashmap (they keys are the names, the values are the text the was matched?) and you have the opportunity to inspect it and decide if it's OK for you.

Correct!

>this requires normal name matchers to be used first, the output of which is piped into the input of name matching processor.

Yes. I should have added a check for the "name matching processor requires a name matcher". Fixed in 4b8ad4f.

> Would name matcher validator be a better name?

That is less bad. Fixed in 03b3d27.

[I think *all* of the names in fm could do with a rethink. "Name and pattern" are easily confused, for example.]

>is the validator called once for each matcher, or with the union of all matchers

Backtracking makes it pretty much impossible to give a simple bound here, unfortunately.